### ISSUE TITLE
segmentation fault with `open_mfdataset`

### ISSUE THREAD

**Issue by razvanc87** [xray/xray#444]
This is super strange. Does anyone have any idea why a segmentation fault might be happening here?

```
Python 3.4.3 (default, Jun 26 2015, 00:02:21) 
[GCC 4.3.4 [gcc-4_3-branch revision 152973]] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import xray
>>> xray.open_mfdataset('2*.nc', concat_dim='time')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/ichec/home/users/razvan/.local/lib/python3.4/site-packages/xray/backends/api.py", line 205, in open_mfdataset
Segmentation fault (core dumped)
```

I stay it's strange because I ended up tracking down the bug to `xray.core.ops.array_equiv`. I have no idea what's going on, but by mistake I found out that if I introduce `isnull(arr1 & arr2)` just before the `return` statement then I don't get the error any more... So my `xray.core.ops.array_equiv` is now:

```
def array_equiv(arr1, arr2):
    """Like np.array_equal, but also allows values to be NaN in both arrays
    """
    arr1, arr2 = as_like_arrays(arr1, arr2)
    if arr1.shape != arr2.shape:
        return False
    # segmentation fault if we don't call this here...
    isnull(arr1 & arr2)
    return bool(((arr1 == arr2) | (isnull(arr1) & isnull(arr2))).all())
```
Thanks...

**Comment by shoyer** [comment]
Oh my, that's bad!

Can you experiment with the `engine` argument to `open_mfdataset` and see if that changes things? For example, try `engine='scipy'` (if this is a netcdf3 files) and `engine='netcdf4'`.

It would be also be helpful to report the dtypes of the arrays that trigger failure in `array_equiv`.

**Comment by razvanc87** [comment]
Unfortunately I can't use `engine='scipy'` cause they're not netcdf3 files so it defaults to `'netcdf4'`. On the other hand [here](https://gist.github.com/razvanc87/0986c4f7a591772e1778) you can find the back trace from `gdb`... if that helps in any way...

```
print(arr1.dtype, arr2.dtype)
print((arr1 == arr2))
print((arr1 == arr2) | (isnull(arr1) & isnull(arr2)))

# gives:
float64 float64
dask.array<x_1, shape=(50, 39, 59), chunks=((50,), (39,), (59,)), dtype=bool>
dask.array<x_6, shape=(50, 39, 59), chunks=((50,), (39,), (59,)), dtype=bool>
```

Funny thing is when I'm adding these print statements and so on I get some traceback from Python (some times). Without them I would only get segmetation fault with no additional information. For example, just now, after introducing these `print`s I got [this](https://gist.github.com/razvanc87/82dc9635f89b55ffaf46) traceback. This doesn't seem to be an `xray` bug, I mean it can't since it's just `Python` code... but any help is appreciated. Thanks!

**Comment by shoyer** [comment]
Another backend to try would be `engine='h5netcdf'`: https://github.com/shoyer/h5netcdf

That might help us identify if this is a netCDF4-python bug.

I am also baffled by how inserting `isnull(arr1 & arr2)` avoids the seg fault. This is a lazy computation created with dask that is immediately thrown away without accessing any of the values.

**Comment by razvanc87** [comment]
Just tried `engine='h5netcdf'`. Still get the segfault. It looks to me that something doesn't properly initialize the hdf5 library and calling that `isnull` function like this somehow triggers some initialization for the both arrays. It might also be the `&` operator... because if I do `isnull(arr1) & isnull(arr2)` I still get the segmentation fault. Only when using `isnull(arr1 & arr2)` it seems to work... strange things.

**Comment by shoyer** [comment]
do you have an example file? this might also be your HDF5 install....

**Comment by mrocklin** [comment]
@shoyer asked me to chime in in case this is an issue with dask.  One thing to try would be to remove multi-threading from the equation.  I'm not sure how this would affect things but it's worth a shot.

```python
>>> import dask
>>> from dask.async import get_sync
>>> dask.set_options(get=get_sync)  # use single-threaded scheduler by default
>>> ... do work as normal
```

**Comment by mrocklin** [comment]
Alternatively can we try doing the operations that xray would do manually and see if one of them triggers something?  

One could also try

    $ gdb python

**Comment by razvanc87** [comment]
So I just tried @mrocklin's idea with using single-threaded stuff. This seems to fix the segmentation fault, but I am very curious as to why there's a problem with working in parallel. I tried two different hdf5 libraries (I think version 1.8.13 and 1.8.14) but I got the same segmentation fault. Anyway, working on a single thread is not a big deal, I'll just do that for the time being... I already tried `gdb` on python but I'm not experienced enough to make heads or tails of it... I have the `gdb` backtrace [here](https://gist.github.com/razvanc87/0986c4f7a591772e1778) but I don't know what to do with it...

@shoyer, the files are not the issue here, they're the same ones I provided in #443.

Question: does the hdf5 library need to be built with parallel support (mpi or something) maybe?... thanks guys

**Comment by mrocklin** [comment]
There was a similar problem with PyTables, which didn't support concurrency
well.  This resulted in the from-hdf5 function in dask array which uses
explicit locks to avoid concurrent access.

We could repeat this treatment more generally without much trouble to force
single threaded access on access but still allow parallelism otherwise.

**Comment by shoyer** [comment]
Of course, concurrent access to HDF5 files works fine on my laptop, using Anaconda's build of HDF5 (version 1.8.14). I have no idea what special flags they invoked when building it :).

That said, I have been unable to produce any benchmarks that show improved performance when simply doing multithreaded *reads* without doing any computation (e.g., `%time xray.open_dataset(..., chunks=...).load()`). Even when I'm reading multiple independent chunks compressed on disk, CPU seems to be pegged at 100%, when using either netCDF4-python or h5py (via h5netcdf) to read the data. For non-compressed data, reads seem to be limited by disk speed, so CPU is also not relevant.

Given these considerations, it seems like we should use a lock when reading data into xray with dask. @mrocklin we could just use `lock=True` with `da.from_array`, right? If we can find use cases for multi-threaded reads, we could also add an optional `lock` argument to `open_dataset`/`open_mfdataset`.

**Comment by mrocklin** [comment]
Oh, I didn't realize that that was built in already.  Sounds like you could
handle this easily on the xray side.

**Comment by shoyer** [comment]
I have a tentative fix (adding the threading lock) in https://github.com/xray/xray/pull/446

Still wondering why multi-threading can't use more than one CPU -- hopefully my h5py issue (referenced above) will get us some answers.

**Comment by shoyer** [comment]
Just merged the fix to master.

@razvanc87 if you could try installing the development version, I would love to hear if this resolves your issues.

**Comment by shoyer** [comment]
@razvanc87 What version of h5py were you using with h5netcdf? @andrewcollette suggests (https://github.com/h5py/h5py/issues/591#issuecomment-116785660) that h5py should already have the lock that fixes this issue if you were using h5py 2.4.0 or later.

**Comment by razvanc87** [comment]
Well... I have a couple of remarks to make. After some more thought about this it might have been all along my fault. Let me explain. I have this machine at work where I don't have administrative privileges so I decided to give `linuxbrew` a try. Now there are some system `hdf5` libraries (but in custom locations) and they have this `module` command to load different versions of packages and set up proper environment variables. Before I had this issue, I did have `xray` installed with `dask` and everything compiled against the system libraries (and I had no problems with it). Then, with `linuxbrew` I started getting this weird behavior, using the latest version of `hdf5` (1.8.14), but then I tried with version (1.8.13) and I had the same issue. Then I read somewhere on the net that... because of this mixture of local - system install with `linuxbrew` there might be issues when compiling, that is, the compiler uses versions of some header files that don't necessarily match local installed libraries. I can't confirm this any more though cause I reconfigured everything and removed `linuxbrew` cause it was producing more problems than solving... but I'll be happy to give the current installation a try and see if I can reproduce the error... can't do more than this though... sorry.

**Comment by razvanc87** [comment]
OK... as a follow-up, I did some tests and with `netcdf4` I got this error again, but using `open_mfdataset` with the latest versions of `h5py` & `h5netcdf` I don't. But there are some decodings that aren't happening now... for whatever reason (maybe `h5netcdf`?). Anyway, my netcdf files store the attributes in `'ascii'`, that is, `bytes` in python so when trying to check for the time I get:
```
Traceback (most recent call last):
  File "segfault.py", line 62, in <module>
    concat_dim='time', engine='h5netcdf'))
  File "/ichec/home/users/razvan/.local/lib/python3.4/site-packages/xray/backends/api.py", line 202, in open_mfdataset
    datasets = [open_dataset(p, **kwargs) for p in paths]
  File "/ichec/home/users/razvan/.local/lib/python3.4/site-packages/xray/backends/api.py", line 202, in <listcomp>
    datasets = [open_dataset(p, **kwargs) for p in paths]
  File "/ichec/home/users/razvan/.local/lib/python3.4/site-packages/xray/backends/api.py", line 145, in open_dataset
    return maybe_decode_store(store)
  File "/ichec/home/users/razvan/.local/lib/python3.4/site-packages/xray/backends/api.py", line 101, in maybe_decode_store
    concat_characters=concat_characters, decode_coords=decode_coords)
  File "/ichec/home/users/razvan/.local/lib/python3.4/site-packages/xray/conventions.py", line 850, in decode_cf
    decode_coords)
  File "/ichec/home/users/razvan/.local/lib/python3.4/site-packages/xray/conventions.py", line 791, in decode_cf_variables
    decode_times=decode_times)
  File "/ichec/home/users/razvan/.local/lib/python3.4/site-packages/xray/conventions.py", line 735, in decode_cf_variable
    if 'since' in attributes['units']:
TypeError: Type str doesn't support the buffer API
```
This is simple to solve.. just have every `byte` attribute decode to `'utf8'` when first reading in the variables... I'll have some more time to look at this alter today.

**Comment by shoyer** [comment]
Thanks for your help debugging!

I made a new issue for ascii attributes handling: https://github.com/xray/xray/issues/451

This is one case where Python 3's insistence that bytes and strings are different is annoying. I'll probably have to decode all bytes type attributes read from h5netcdf.

How do you trigger the seg-fault with netcdf4-python? Just using `open_mfdataset` as before? I'm a little surprised that still happens with the thread lock.

**Comment by razvanc87** [comment]
Yes, I'm using the same files that I [once uploaded on Dropbox](https://www.dropbox.com/sh/wi7s59wu1soh5f3/AABxHjZ1ssXzIeAMwLG5cKAea?dl=0) for you to play with for #443. I'm not doing anything special, just passing in the glob pattern to `open_mfdataset` with no option for `engine` (which I guess goes for `netcdf4` by default).

**Comment by shoyer** [comment]
Ah, I think I know why the seg faults are still occuring. By default, `dask.array.from_array` uses a thread lock that is specific to each array variable. We need a global thread lock, because the HDF5 library is not thread safe.

@mrocklin maybe `da.from_array` should use a global thread lock if `lock=True`? Alternatively, I could just change this in xray -- but I suspect that other dask users who want a lock also probably want a global lock.

**Comment by mrocklin** [comment]
The library itself is not threadsafe?  What about on a per-file basis?

**Comment by razvanc87** [comment]
Per file basis (`open_dataset`) there's no problem... but again, if I try `h5netcdf` engine, `open_mfdataset` doesn't throw a segmentation fault, but then I go into the string unicode/ascii problem. So I guess `h5netcdf` and `netcdf4` use the same netcdf/hdf5 libraries don't they? so if if works for `h5netcdf` then it should work for `netcdf4` as well...

**Comment by shoyer** [comment]
> The library itself is not threadsafe? What about on a per-file basis?

@andrewcollette could you comment on this for h5py/hdf5?

@mrocklin based on my reading of Andrew's comment in the h5py issue, this is indeed the case.

**Comment by shoyer** [comment]
@razvanc87 netcdf4 and h5py use the same HDF5 libraries, but have different bindings from Python. H5py likely does a more careful job of using its own locks to ensure thread safety, which likely explains the difference you are seeing (the attribute encoding is a separate issue).

**Comment by andrewcollette** [comment]
@shoyer, there are basically two levels of thread safety for HDF5/h5py. First, the HDF5 library has an optional compile-time "threadsafe" build option that wraps all API access in a lock.  This is all-or-nothing; I'm not aware of any per-file effects.

Second, h5py uses its own global lock on the Python side to serialize access, which is only disabled in MPI mode.  For added protection, h5py also does not presently release the GIL around reads/writes.

**Comment by razvanc87** [comment]
I think this issue can be closed, after some digging and playing with different `netcdf4` modules I'm pretty certain that it was a linkage and compilation issue between system `hdf5` and `netcdf` libraries. You see, the computer I got this error on is one of those "module load" managed supercomputers... and somewhere on the way things got messed up while compiling python modules...

**Comment by shoyer** [comment]
@razvanc87 I've gotten a few other reports of issues with multithreading (not just you), so I think we do definitely need to add our own lock when accessing these files. Misconfigured hdf5 installs may not be so  uncommon.
